client/checker_client/bigmap.py

```diff
@@ -0,0 +1,162 @@
+"""Utilities for working with contracts which use big_maps
+"""
+
+from typing import Any, Dict, Optional, Set
+
+from pytezos.client import PyTezosClient
+from pytezos.contract.interface import ContractInterface
+from pytezos.michelson.micheline import micheline_value_to_python_object
+from pytezos.michelson.types import ListType
+from pytezos.michelson.types.adt import ADTMixin
+from pytezos.michelson.types.base import Undefined
+from pytezos.michelson.types.big_map import BigMapType
+from pytezos.rpc.search import BlockSliceQuery
+
+
+def scan_bigmap_keys(
+    tz: PyTezosClient,
+    contract_address: str,
+    start_block: int = 1,
+    end_after_blocks: Optional[int] = None,
+) -> Dict[str, Set[Any]]:
+    """Scans for all big_map keys in a contract's storage
+
+    This function works by scanning the main chain for any operations on
+    the specified contract which produce a lazy_storage_diff.
+
+    Args:
+        tz: pytezos client instance
+        contract_address: The address of the contract of interest
+        start_block: The block at which to start scanning. Note that setting this to a block after
+            the contract's origination may cause you to miss keys.
+        end_after_blocks: Optionally end the scan after this many blocks without an update
+            to a big_map. The default value of None will cause the function to scan the entire
+            chain.
+    Returns:
+        A dict mapping bigmap_ids to their keys. Note that ids may be missing from here if no
+            keys are ever added to the big_map (i.e. it is only allocated).
+    """
+    diffs = []
+    blocks_since_update = None
+    blocks: BlockSliceQuery = tz.shell.blocks[start_block:]
+    # First, need to collect the lazy storage diffs
+    for b in blocks:
+        block_diffs = []
+        if contract_address in b.context.contracts():
+            # Contract exists in this operation. Let's look for operations on it
+            ops = tz.shell.blocks[b.hash()].operations.managers()
+            for op_group in ops:
+                op_group = op_group["contents"]
+                for op in op_group:
+                    op_result = op["metadata"]["operation_result"]
+                    # Skip ops which don't operate on the contract of interest
+                    if (
+                        op["kind"] == "origination"
+                        and contract_address not in op_result["originated_contracts"]
+                    ):
+                        continue
+                    if op["kind"] == "transaction" and op["destination"] != contract_address:
+                        continue
```

Comment:
Do we want to check `op["source"]` here too, perhaps? Maybe not, since checker is originated :thinking:

Reply:
I think it is okay to leave out in this case. With the current implementation we also collect big_map diffs which occurred due to transactions besides those we sent, which I think might actually be desirable when using this outside of the sandbox environment.